commit 1285be2d9f46ae2b66034f443d6d75857f93673a
Author: Vladimir Zdorovenco <vr.zdor@gmail.com>
Date:   Mon Jan 10 23:25:45 2022 +0200

    initial.

diff --git a/arch/mips/Kconfig b/arch/mips/Kconfig
index 0c21593a8..1771e8a6e 100644
--- a/arch/mips/Kconfig
+++ b/arch/mips/Kconfig
@@ -640,6 +640,8 @@ config MIPS_TC3262
 	select SYS_SUPPORTS_BIG_ENDIAN
 	select SYS_SUPPORTS_LITTLE_ENDIAN
 	select SYS_SUPPORTS_MULTITHREADING
+	select USE_OF
+	select RESET_CONTROLLER
 	# select TC3162_IMEM
 	# select TC3162_DMEM
 	help
diff --git a/arch/mips/tc3162/irq.c b/arch/mips/tc3162/irq.c
index 008380908..dacebec7d 100644
--- a/arch/mips/tc3162/irq.c
+++ b/arch/mips/tc3162/irq.c
@@ -513,6 +513,7 @@ void __init arch_init_irq(void)
 #ifdef CONFIG_MIPS_MT_SMP
 	vsmp_int_init();
 #endif
+	tc3162_enable_irq(MAC_INT);
 }
 
 asmlinkage void plat_irq_dispatch(void)
diff --git a/arch/mips/tc3162/prom.c b/arch/mips/tc3162/prom.c
index c9e34be72..407c970e3 100644
--- a/arch/mips/tc3162/prom.c
+++ b/arch/mips/tc3162/prom.c
@@ -1,9 +1,8 @@
 #include <linux/init.h>
 #include <linux/mm.h>
 #include <linux/sched.h>
-// TODO: No such file.
-// #include <linux/bootmem.h>
 #include <linux/blkdev.h>
+#include <linux/of_fdt.h>
 
 #include <asm/mipsmtregs.h>
 #include <asm/addrspace.h>
@@ -12,6 +11,7 @@
 #include <asm/time.h>
 #include <asm/tc3162/tc3162.h>
 #include <asm/traps.h>
+#include <asm/prom.h>
 
 extern int __imem, __dmem;
 
@@ -323,6 +323,16 @@ void __init prom_init(void)
 		cpu_dma_round_robin(ENABLE);
 }
 
+void __init device_tree_init(void)
+{
+	/* fw_passed_dtb will be set if DTB is added to the 
+	   end of vmlinux.bin, see MIPS_RAW_APPENDED_DTB. */
+	if (fw_passed_dtb) {
+		__dt_setup_arch(fw_passed_dtb);
+		unflatten_and_copy_device_tree();
+	}
+}
+
 void __init prom_free_prom_memory(void)
 {
 	/* We do not have any memory to free */
diff --git a/drivers/mfd/syscon.c b/drivers/mfd/syscon.c
index ca465794e..e3fbaa9a8 100644
--- a/drivers/mfd/syscon.c
+++ b/drivers/mfd/syscon.c
@@ -60,7 +60,10 @@ static struct syscon *of_syscon_register(struct device_node *np, bool check_clk)
 		goto err_map;
 	}
 
-	base = ioremap(res.start, resource_size(&res));
+	
+	/* FIXME: With ioremap, regmap_ calls will wait forever. */
+	/* mips/ralinks calls: set_io_port_base(). */
+	base = res.start; // ioremap(res.start, resource_size(&res));
 	if (!base) {
 		ret = -ENOMEM;
 		goto err_map;
diff --git a/drivers/net/ethernet/mediatek/Kconfig b/drivers/net/ethernet/mediatek/Kconfig
index 3362b148d..1a9b973f4 100644
--- a/drivers/net/ethernet/mediatek/Kconfig
+++ b/drivers/net/ethernet/mediatek/Kconfig
@@ -1,7 +1,7 @@
 # SPDX-License-Identifier: GPL-2.0-only
 config NET_VENDOR_MEDIATEK
 	bool "MediaTek devices"
-	depends on ARCH_MEDIATEK || SOC_MT7621 || SOC_MT7620
+	depends on ARCH_MEDIATEK || SOC_MT7621 || SOC_MT7620 || MIPS_TC3262 
 	help
 	  If you have a Mediatek SoC with ethernet, say Y.
 
