diff -r -I '$Id:' -I '$Header' -I '$Revision' -I printk -I VERSION -I '^\*' -x '*.lds' -x '*.order' -x '*.builtin' -x '*.a' -x '*.d' -x '*.o' -x '*.cmd' -x '*.ko' -x '*.[chS]~' -x cevt-smtc.c -x smtc.c -x '*ecnt*' -x pci-7512api.c -x softfloat -x '*usb*' -x '*ralink*' -x .gitignore -ur linux-2.6.36/arch/mips/include/asm/cpu.h vr300-linux/arch/mips/include/asm/cpu.h
--- linux-2.6.36/arch/mips/include/asm/cpu.h	2010-10-20 23:30:22.000000000 +0300
+++ vr300-linux/arch/mips/include/asm/cpu.h	2019-03-07 03:40:17.000000000 +0200
@@ -72,6 +72,8 @@
 #define PRID_IMP_R5500		0x5500
 #define PRID_IMP_LOONGSON2	0x6300
 
+#define PRID_IMP_TC3162		0xcd00
+
 #define PRID_IMP_UNKNOWN	0xff00
 
 /*
diff -r -I '$Id:' -I '$Header' -I '$Revision' -I printk -I VERSION -I '^\*' -x '*.lds' -x '*.order' -x '*.builtin' -x '*.a' -x '*.d' -x '*.o' -x '*.cmd' -x '*.ko' -x '*.[chS]~' -x cevt-smtc.c -x smtc.c -x '*ecnt*' -x pci-7512api.c -x softfloat -x '*usb*' -x '*ralink*' -x .gitignore -ur linux-2.6.36/arch/mips/include/asm/spram.h vr300-linux/arch/mips/include/asm/spram.h
--- linux-2.6.36/arch/mips/include/asm/spram.h	2010-10-20 23:30:22.000000000 +0300
+++ vr300-linux/arch/mips/include/asm/spram.h	2019-03-07 03:40:17.000000000 +0200
@@ -7,4 +7,8 @@
 static inline void spram_config(void) { };
 #endif /* CONFIG_CPU_MIPSR2 */
 
+extern int is_sram_addr(void *p);
+extern void *alloc_sram(int n);
+extern void free_sram(void *p, int n);
+
 #endif /* _MIPS_SPRAM_H */
Only in vr300-linux/arch/mips/include/asm: tc3162
diff -r -I '$Id:' -I '$Header' -I '$Revision' -I printk -I VERSION -I '^\*' -x '*.lds' -x '*.order' -x '*.builtin' -x '*.a' -x '*.d' -x '*.o' -x '*.cmd' -x '*.ko' -x '*.[chS]~' -x cevt-smtc.c -x smtc.c -x '*ecnt*' -x pci-7512api.c -x softfloat -x '*usb*' -x '*ralink*' -x .gitignore -ur linux-2.6.36/arch/mips/include/asm/thread_info.h vr300-linux/arch/mips/include/asm/thread_info.h
--- linux-2.6.36/arch/mips/include/asm/thread_info.h	2010-10-20 23:30:22.000000000 +0300
+++ vr300-linux/arch/mips/include/asm/thread_info.h	2021-12-10 01:44:56.979644988 +0200
@@ -62,8 +62,12 @@
 
 /* thread information allocation */
 #if defined(CONFIG_PAGE_SIZE_4KB) && defined(CONFIG_32BIT)
+#if defined(CONFIG_CPU_TC3162) || defined(CONFIG_MIPS_TC3262)
+#define THREAD_SIZE_ORDER (2)
+#else
 #define THREAD_SIZE_ORDER (1)
 #endif
+#endif
 #if defined(CONFIG_PAGE_SIZE_4KB) && defined(CONFIG_64BIT)
 #define THREAD_SIZE_ORDER (2)
 #endif
diff -r -I '$Id:' -I '$Header' -I '$Revision' -I printk -I VERSION -I '^\*' -x '*.lds' -x '*.order' -x '*.builtin' -x '*.a' -x '*.d' -x '*.o' -x '*.cmd' -x '*.ko' -x '*.[chS]~' -x cevt-smtc.c -x smtc.c -x '*ecnt*' -x pci-7512api.c -x softfloat -x '*usb*' -x '*ralink*' -x .gitignore -ur linux-2.6.36/arch/mips/Kbuild.platforms vr300-linux/arch/mips/Kbuild.platforms
--- linux-2.6.36/arch/mips/Kbuild.platforms	2010-10-20 23:30:22.000000000 +0300
+++ vr300-linux/arch/mips/Kbuild.platforms	2021-12-10 02:07:20.506003560 +0200
@@ -18,6 +18,7 @@
 platforms += pnx833x
 platforms += pnx8550
 platforms += powertv
+platforms += ralink
 platforms += rb532
 platforms += sgi-ip22
 platforms += sgi-ip27
diff -r -I '$Id:' -I '$Header' -I '$Revision' -I printk -I VERSION -I '^\*' -x '*.lds' -x '*.order' -x '*.builtin' -x '*.a' -x '*.d' -x '*.o' -x '*.cmd' -x '*.ko' -x '*.[chS]~' -x cevt-smtc.c -x smtc.c -x '*ecnt*' -x pci-7512api.c -x softfloat -x '*usb*' -x '*ralink*' -x .gitignore -ur linux-2.6.36/arch/mips/Kconfig vr300-linux/arch/mips/Kconfig
--- linux-2.6.36/arch/mips/Kconfig	2010-10-20 23:30:22.000000000 +0300
+++ vr300-linux/arch/mips/Kconfig	2019-03-07 03:40:18.000000000 +0200
@@ -252,6 +252,65 @@
 	  This enables support for the MIPS Technologies Malta evaluation
 	  board.
 
+config MIPS_TC3262
+	bool "TrendChip's TC3262 Board"
+	select CPU_MIPSR2_IRQ_VI
+	select CPU_MIPSR2_IRQ_EI
+	select NO_EXCEPT_FILL
+	select BOOT_RAW
+	select CEVT_R4K
+	select CSRC_R4K
+	select DMA_NONCOHERENT
+	select IRQ_CPU
+	select HW_HAS_PCI
+	select SWAP_IO_SPACE
+	select SYS_HAS_CPU_MIPS32_R1
+	select SYS_HAS_CPU_MIPS32_R2
+	select SYS_HAS_CPU_MIPS64_R1
+	select SYS_HAS_EARLY_PRINTK
+	select SYS_SUPPORTS_32BIT_KERNEL
+	select SYS_SUPPORTS_BIG_ENDIAN
+	select SYS_SUPPORTS_LITTLE_ENDIAN
+	select SYS_SUPPORTS_MULTITHREADING
+	select TC3162_ADSL
+	select TC3162_IMEM
+	select TC3162_DMEM
+	help
+	  This enables support for TrendChip's TC3262 based board.  
+	  board.
+
+config MIPS_TC3182
+	bool "TrendChip's TC3182 Board"
+	depends on MIPS_TC3262
+	help
+	  This enables support for TrendChip's TC3182 based board.  
+	  board.
+
+config MIPS_RT63165
+	bool "Ralink's RT63165 Board"
+	depends on MIPS_TC3262
+	help
+	  This enables support for Ralink's RT63165 based board.  
+
+config MIPS_RT65168
+	bool "Ralink's RT65168 Board"
+	depends on MIPS_TC3262
+	select RALINK_VDSL
+	help
+	  This enables support for Ralink's RT65168 based board.  
+
+config MIPS_RT63365
+	bool "Ralink's RT63365 Board"
+	depends on MIPS_TC3262
+	help
+	  This enables support for Ralink's RT63365 based board.  
+config MIPS_MT7510
+	bool "MediaTek's MT7510 Board"
+	depends on MIPS_TC3262
+	select RALINK_VDSL
+	help
+	  This enables support for MediaTek's MT7510 based board.  
+	  
 config MIPS_SIM
 	bool 'MIPS simulator (MIPSsim)'
 	select CEVT_R4K
@@ -1053,6 +1112,32 @@
 config BOOT_ELF32
 	bool
 
+config MIPS_PATENTFREE
+	bool
+	depends on SYS_HAS_CPU_TC3162
+
+config TC3162_ADSL
+	bool 
+	depends on SYS_HAS_CPU_TC3162
+
+config RALINK_VDSL
+	bool 
+	depends on SYS_HAS_CPU_TC3162
+
+config TC3162_IMEM
+	bool 
+	depends on SYS_HAS_CPU_TC3162
+
+config TC3162_DMEM
+	bool 
+	depends on SYS_HAS_CPU_TC3162
+
+config IMEM_SIZE
+	int
+	default "32768" if MIPS_TC3162U || MIPS_TC3182 || MIPS_TC3262
+	default "16384" if MIPS_TC3162
+	depends on TC3162_IMEM
+
 config MIPS_L1_CACHE_SHIFT
 	int
 	default "4" if MACH_DECSTATION || MIKROTIK_RB532 || PMC_MSP4200_EVAL
@@ -2303,3 +2388,5 @@
 source "crypto/Kconfig"
 
 source "lib/Kconfig"
+
+source "pwModel/Kconfig"
Only in vr300-linux/arch/mips/kernel: asm-offsets.s
diff -r -I '$Id:' -I '$Header' -I '$Revision' -I printk -I VERSION -I '^\*' -x '*.lds' -x '*.order' -x '*.builtin' -x '*.a' -x '*.d' -x '*.o' -x '*.cmd' -x '*.ko' -x '*.[chS]~' -x cevt-smtc.c -x smtc.c -x '*ecnt*' -x pci-7512api.c -x softfloat -x '*usb*' -x '*ralink*' -x .gitignore -ur linux-2.6.36/arch/mips/kernel/cpu-probe.c vr300-linux/arch/mips/kernel/cpu-probe.c
--- linux-2.6.36/arch/mips/kernel/cpu-probe.c	2010-10-20 23:30:22.000000000 +0300
+++ vr300-linux/arch/mips/kernel/cpu-probe.c	2021-12-06 09:21:25.378872572 +0200
@@ -356,6 +356,13 @@
 			c->options |= MIPS_CPU_FPU;
 		c->tlbsize = 64;
 		break;
+	case PRID_IMP_TC3162:
+		c->cputype = CPU_R3000;
+		c->isa_level = MIPS_CPU_ISA_I;
+		c->options = MIPS_CPU_TLB | MIPS_CPU_3K_CACHE |
+		             MIPS_CPU_NOFPUEX;
+		c->tlbsize = 32;
+		break;
 	case PRID_IMP_R4000:
 		if (read_c0_config() & CONF_SC) {
 			if ((c->processor_id & 0xff) >= PRID_REV_R4400) {
diff -r -I '$Id:' -I '$Header' -I '$Revision' -I printk -I VERSION -I '^\*' -x '*.lds' -x '*.order' -x '*.builtin' -x '*.a' -x '*.d' -x '*.o' -x '*.cmd' -x '*.ko' -x '*.[chS]~' -x cevt-smtc.c -x smtc.c -x '*ecnt*' -x pci-7512api.c -x softfloat -x '*usb*' -x '*ralink*' -x .gitignore -ur linux-2.6.36/arch/mips/kernel/csrc-r4k.c vr300-linux/arch/mips/kernel/csrc-r4k.c
--- linux-2.6.36/arch/mips/kernel/csrc-r4k.c	2010-10-20 23:30:22.000000000 +0300
+++ vr300-linux/arch/mips/kernel/csrc-r4k.c	2019-03-07 03:40:15.000000000 +0200
@@ -15,7 +15,8 @@
 	return read_c0_count();
 }
 
-static struct clocksource clocksource_mips = {
+//static 
+struct clocksource clocksource_mips = {
 	.name		= "MIPS",
 	.read		= c0_hpt_read,
 	.mask		= CLOCKSOURCE_MASK(32),
diff -r -I '$Id:' -I '$Header' -I '$Revision' -I printk -I VERSION -I '^\*' -x '*.lds' -x '*.order' -x '*.builtin' -x '*.a' -x '*.d' -x '*.o' -x '*.cmd' -x '*.ko' -x '*.[chS]~' -x cevt-smtc.c -x smtc.c -x '*ecnt*' -x pci-7512api.c -x softfloat -x '*usb*' -x '*ralink*' -x .gitignore -ur linux-2.6.36/arch/mips/kernel/entry.S vr300-linux/arch/mips/kernel/entry.S
--- linux-2.6.36/arch/mips/kernel/entry.S	2010-10-20 23:30:22.000000000 +0300
+++ vr300-linux/arch/mips/kernel/entry.S	2019-03-07 03:40:15.000000000 +0200
@@ -123,7 +123,7 @@
 	SAVE_AT
 	SAVE_TEMP
 	LONG_L	v0, PT_STATUS(sp)
-#if defined(CONFIG_CPU_R3000) || defined(CONFIG_CPU_TX39XX)
+#if defined(CONFIG_CPU_R3000) || defined(CONFIG_CPU_TX39XX) || defined(CONFIG_CPU_TC3162)
 	and	v0, ST0_IEP
 #else
 	and	v0, ST0_IE
diff -r -I '$Id:' -I '$Header' -I '$Revision' -I printk -I VERSION -I '^\*' -x '*.lds' -x '*.order' -x '*.builtin' -x '*.a' -x '*.d' -x '*.o' -x '*.cmd' -x '*.ko' -x '*.[chS]~' -x cevt-smtc.c -x smtc.c -x '*ecnt*' -x pci-7512api.c -x softfloat -x '*usb*' -x '*ralink*' -x .gitignore -ur linux-2.6.36/arch/mips/kernel/genex.S vr300-linux/arch/mips/kernel/genex.S
--- linux-2.6.36/arch/mips/kernel/genex.S	2010-10-20 23:30:22.000000000 +0300
+++ vr300-linux/arch/mips/kernel/genex.S	2021-12-06 12:04:59.842775420 +0200
@@ -178,7 +178,7 @@
 	.set	push
 	.set	noat
 	mfc0	k0, CP0_STATUS
-#if defined(CONFIG_CPU_R3000) || defined(CONFIG_CPU_TX39XX)
+#if defined(CONFIG_CPU_R3000) || defined(CONFIG_CPU_TX39XX) || defined(CONFIG_CPU_TC3162)
 	and	k0, ST0_IEP
 	bnez	k0, 1f
 
@@ -512,7 +512,7 @@
 	get_saved_sp	/* k1 := current_thread_info */
 	.set	noreorder
 	MFC0	k0, CP0_EPC
-#if defined(CONFIG_CPU_R3000) || defined(CONFIG_CPU_TX39XX)
+#if defined(CONFIG_CPU_R3000) || defined(CONFIG_CPU_TX39XX) || defined(CONFIG_CPU_TC3162)
 	ori	k1, _THREAD_MASK
 	xori	k1, _THREAD_MASK
 	LONG_L	v1, TI_TP_VALUE(k1)
diff -r -I '$Id:' -I '$Header' -I '$Revision' -I printk -I VERSION -I '^\*' -x '*.lds' -x '*.order' -x '*.builtin' -x '*.a' -x '*.d' -x '*.o' -x '*.cmd' -x '*.ko' -x '*.[chS]~' -x cevt-smtc.c -x smtc.c -x '*ecnt*' -x pci-7512api.c -x softfloat -x '*usb*' -x '*ralink*' -x .gitignore -ur linux-2.6.36/arch/mips/kernel/Makefile vr300-linux/arch/mips/kernel/Makefile
--- linux-2.6.36/arch/mips/kernel/Makefile	2010-10-20 23:30:22.000000000 +0300
+++ vr300-linux/arch/mips/kernel/Makefile	2019-03-07 03:40:15.000000000 +0200
@@ -36,6 +36,7 @@
 obj-$(CONFIG_CPU_MIPS32)	+= r4k_fpu.o r4k_switch.o
 obj-$(CONFIG_CPU_MIPS64)	+= r4k_fpu.o r4k_switch.o
 obj-$(CONFIG_CPU_R3000)		+= r2300_fpu.o r2300_switch.o
+obj-$(CONFIG_CPU_TC3162)	+= r2300_fpu.o r2300_switch.o
 obj-$(CONFIG_CPU_R4300)		+= r4k_fpu.o r4k_switch.o
 obj-$(CONFIG_CPU_R4X00)		+= r4k_fpu.o r4k_switch.o
 obj-$(CONFIG_CPU_R5000)		+= r4k_fpu.o r4k_switch.o
diff -r -I '$Id:' -I '$Header' -I '$Revision' -I printk -I VERSION -I '^\*' -x '*.lds' -x '*.order' -x '*.builtin' -x '*.a' -x '*.d' -x '*.o' -x '*.cmd' -x '*.ko' -x '*.[chS]~' -x cevt-smtc.c -x smtc.c -x '*ecnt*' -x pci-7512api.c -x softfloat -x '*usb*' -x '*ralink*' -x .gitignore -ur linux-2.6.36/arch/mips/kernel/mips-mt-fpaff.c vr300-linux/arch/mips/kernel/mips-mt-fpaff.c
--- linux-2.6.36/arch/mips/kernel/mips-mt-fpaff.c	2010-10-20 23:30:22.000000000 +0300
+++ vr300-linux/arch/mips/kernel/mips-mt-fpaff.c	2019-03-07 03:40:15.000000000 +0200
@@ -73,6 +73,28 @@
 	if (copy_from_user(&new_mask, user_mask_ptr, sizeof(new_mask)))
 		return -EFAULT;
 
+
+#ifdef CONFIG_MIPS_MT_SMP
+#ifndef CONFIG_MIPS_MT_SMTC
+	printk("\r\n\r\n%s new_mask origin value is %08x\r\n\r\n", __FUNCTION__, *new_mask);
+	if(*(unsigned long *)new_mask == 0x8)
+	{
+		*(unsigned long *)new_mask = 0x2;
+	}
+	else if(*(unsigned long *)new_mask == 0xf)
+	{
+		*(unsigned long *)new_mask = 0x3;
+	}
+	else
+	{
+		*(unsigned long *)new_mask = 0x1;
+	}
+	
+	printk("\r\n\r\n%s new_mask changed value is %08x\r\n\r\n", __FUNCTION__, *new_mask);
+#endif
+#endif
+
+
 	get_online_cpus();
 	rcu_read_lock();
 
diff -r -I '$Id:' -I '$Header' -I '$Revision' -I printk -I VERSION -I '^\*' -x '*.lds' -x '*.order' -x '*.builtin' -x '*.a' -x '*.d' -x '*.o' -x '*.cmd' -x '*.ko' -x '*.[chS]~' -x cevt-smtc.c -x smtc.c -x '*ecnt*' -x pci-7512api.c -x softfloat -x '*usb*' -x '*ralink*' -x .gitignore -ur linux-2.6.36/arch/mips/kernel/vmlinux.lds.S vr300-linux/arch/mips/kernel/vmlinux.lds.S
--- linux-2.6.36/arch/mips/kernel/vmlinux.lds.S	2010-10-20 23:30:22.000000000 +0300
+++ vr300-linux/arch/mips/kernel/vmlinux.lds.S	2019-03-07 03:40:15.000000000 +0200
@@ -4,6 +4,10 @@
 
 #undef mips
 #define mips mips
+#if defined(TCSUPPORT_CPU_EN7512) || defined(TCSUPPORT_CPU_EN7521)
+#undef CONFIG_IMEM_SIZE
+#define CONFIG_IMEM_SIZE 65536
+#endif
 OUTPUT_ARCH(mips)
 ENTRY(kernel_entry)
 PHDRS {
@@ -51,7 +55,21 @@
 		*(.fixup)
 		*(.gnu.warning)
 	} :text = 0
+
+#ifdef CONFIG_TC3162_IMEM
+#ifdef CONFIG_CPU_TC3162
+  	. = ALIGN(16384);
+#else
+  	. = ALIGN(32768);
+#endif
+  	__imem = . ;
+  	.imem_text : { *(.imem_text) }
+  	_imem_end = .;
+  	_etext = (__imem + CONFIG_IMEM_SIZE) > . ? (__imem + CONFIG_IMEM_SIZE) : .;
+  	. = _etext;
+#else
 	_etext = .;	/* End of text section */
+#endif  
 
 	EXCEPTION_TABLE(16)
 
diff -r -I '$Id:' -I '$Header' -I '$Revision' -I printk -I VERSION -I '^\*' -x '*.lds' -x '*.order' -x '*.builtin' -x '*.a' -x '*.d' -x '*.o' -x '*.cmd' -x '*.ko' -x '*.[chS]~' -x cevt-smtc.c -x smtc.c -x '*ecnt*' -x pci-7512api.c -x softfloat -x '*usb*' -x '*ralink*' -x .gitignore -ur linux-2.6.36/arch/mips/mm/c-r4k.c vr300-linux/arch/mips/mm/c-r4k.c
--- linux-2.6.36/arch/mips/mm/c-r4k.c	2010-10-20 23:30:22.000000000 +0300
+++ vr300-linux/arch/mips/mm/c-r4k.c	2019-03-07 03:40:15.000000000 +0200
@@ -595,7 +595,7 @@
 
 #ifdef CONFIG_DMA_NONCOHERENT
 
-static void r4k_dma_cache_wback_inv(unsigned long addr, unsigned long size)
+__IMEM static void r4k_dma_cache_wback_inv(unsigned long addr, unsigned long size)
 {
 	/* Catch bad driver code */
 	BUG_ON(size == 0);
@@ -623,7 +623,7 @@
 	bc_wback_inv(addr, size);
 }
 
-static void r4k_dma_cache_inv(unsigned long addr, unsigned long size)
+__IMEM static void r4k_dma_cache_inv(unsigned long addr, unsigned long size)
 {
 	/* Catch bad driver code */
 	BUG_ON(size == 0);
Only in vr300-linux/arch/mips/mm: c-tc3162.c
diff -r -I '$Id:' -I '$Header' -I '$Revision' -I printk -I VERSION -I '^\*' -x '*.lds' -x '*.order' -x '*.builtin' -x '*.a' -x '*.d' -x '*.o' -x '*.cmd' -x '*.ko' -x '*.[chS]~' -x cevt-smtc.c -x smtc.c -x '*ecnt*' -x pci-7512api.c -x softfloat -x '*usb*' -x '*ralink*' -x .gitignore -ur linux-2.6.36/arch/mips/mm/Makefile vr300-linux/arch/mips/mm/Makefile
--- linux-2.6.36/arch/mips/mm/Makefile	2010-10-20 23:30:22.000000000 +0300
+++ vr300-linux/arch/mips/mm/Makefile	2019-03-07 03:40:15.000000000 +0200
@@ -16,6 +16,7 @@
 obj-$(CONFIG_CPU_NEVADA)	+= c-r4k.o cex-gen.o tlb-r4k.o
 obj-$(CONFIG_CPU_R10000)	+= c-r4k.o cex-gen.o tlb-r4k.o
 obj-$(CONFIG_CPU_R3000)		+= c-r3k.o tlb-r3k.o
+obj-$(CONFIG_CPU_TC3162)	+= c-tc3162.o tlb-r3k.o pg-r4k.o
 obj-$(CONFIG_CPU_R4300)		+= c-r4k.o cex-gen.o tlb-r4k.o
 obj-$(CONFIG_CPU_R4X00)		+= c-r4k.o cex-gen.o tlb-r4k.o
 obj-$(CONFIG_CPU_R5000)		+= c-r4k.o cex-gen.o tlb-r4k.o
Only in vr300-linux/arch/mips/pci: fixup-tc3162.c
Only in vr300-linux/arch/mips/pci: fixup-tc3162u.c
diff -r -I '$Id:' -I '$Header' -I '$Revision' -I printk -I VERSION -I '^\*' -x '*.lds' -x '*.order' -x '*.builtin' -x '*.a' -x '*.d' -x '*.o' -x '*.cmd' -x '*.ko' -x '*.[chS]~' -x cevt-smtc.c -x smtc.c -x '*ecnt*' -x pci-7512api.c -x softfloat -x '*usb*' -x '*ralink*' -x .gitignore -ur linux-2.6.36/arch/mips/pci/Makefile vr300-linux/arch/mips/pci/Makefile
--- linux-2.6.36/arch/mips/pci/Makefile	2010-10-20 23:30:22.000000000 +0300
+++ vr300-linux/arch/mips/pci/Makefile	2019-03-07 03:40:15.000000000 +0200
@@ -19,6 +19,18 @@
 obj-$(CONFIG_BCM63XX)		+= pci-bcm63xx.o fixup-bcm63xx.o \
 					ops-bcm63xx.o
 
+obj-$(CONFIG_MIPS_TC3162)	+= ops-tc3162.o pci-tc3162.o fixup-tc3162.o
+ifdef CONFIG_MIPS_TC3262 
+EXTRA_CFLAGS += -DPCIE_PCI_COEXIT
+endif
+obj-$(CONFIG_MIPS_TC3262)	+= ops-tc3162u.o pci-tc3162u.o fixup-tc3162u.o pcie-phy.o
+obj-$(CONFIG_MIPS_TC3262)       += ops-tc3162.o pci-tc3162.o fixup-tc3162.o pci-7512api.o
+ifdef CONFIG_MIPS_TC3162U
+EXTRA_CFLAGS += -DPCIE_PCI_COEXIT
+endif
+obj-$(CONFIG_MIPS_TC3162U)	+= ops-tc3162u.o pci-tc3162u.o fixup-tc3162u.o pcie-phy.o
+obj-$(CONFIG_MIPS_TC3162U)	+= ops-tc3162.o pci-tc3162.o fixup-tc3162.o
+
 #
 # These are still pretty much in the old state, watch, go blind.
 #
Only in vr300-linux/arch/mips/pci: ops-tc3162.c
Only in vr300-linux/arch/mips/pci: ops-tc3162u.c
Only in vr300-linux/arch/mips/pci: pcie-phy.c
Only in vr300-linux/arch/mips/pci: pci-tc3162.c
Only in vr300-linux/arch/mips/pci: pci-tc3162u.c
