--- vr300-linux/arch/mips/kernel/cevt-r4k.c	2021-12-06 10:03:59.841363767 +0200
+++ linux-2.6.36/arch/mips/kernel/cevt-r4k.c	2010-10-20 23:30:22.000000000 +0300
@@ -14,7 +14,6 @@
 #include <asm/smtc_ipi.h>
 #include <asm/time.h>
 #include <asm/cevt-r4k.h>
-#include <asm/tc3162/tc3162.h>
 
 /*
  * The SMTC Kernel for the 34K, 1004K, et. al. replaces several
@@ -48,12 +47,12 @@
 int cp0_timer_irq_installed;
 
 #ifndef CONFIG_MIPS_MT_SMTC
+
 irqreturn_t c0_compare_interrupt(int irq, void *dev_id)
 {
 	const int r2 = cpu_has_mips_r2;
 	struct clock_event_device *cd;
 	int cpu = smp_processor_id();
-	unsigned int tmp;
 
 	/*
 	 * Suckage alert:
@@ -64,16 +63,6 @@
 	if (handle_perf_irq(r2))
 		goto out;
 
-	if (isRT63165 || isRT63365 || isMT751020 || isMT7505 || isEN751221) {
-		if (cpu == 0) {
-			mips_timer_ack();
-		}
-		else{
-			tmp = regRead32(CR_CPUTMR_CNT1) + (mips_hpt_frequency/HZ);
-                        regWrite32(CR_CPUTMR_CMR1, tmp);
-		}
-	}
-
 	/*
 	 * The same applies to performance counter interrupts.  But with the
 	 * above we now know that the reason we got here must be a timer
