--- vr300-linux-orig/arch/mips/kernel/genex.S	2021-12-05 22:52:39.547293907 +0200
+++ vr300-linux/arch/mips/kernel/genex.S	2021-12-06 12:04:59.842775420 +0200
@@ -278,7 +278,6 @@
 	 * service routine will have cleared the state, and any active
 	 * level represents a new or otherwised unserviced event...
 	 */
-#ifndef CONFIG_MIPS_TC3262
 	mfc0	t1, CP0_STATUS
 	and	t0, a0, t1
 #ifdef CONFIG_MIPS_MT_SMTC_IM_BACKSTOP
@@ -289,7 +288,6 @@
 	xor	t1, t1, t0
 	mtc0	t1, CP0_STATUS
 	_ehb
-#endif
 #endif /* CONFIG_MIPS_MT_SMTC */
 	CLI
 #ifdef CONFIG_TRACE_IRQFLAGS
