--- vr300-linux/arch/mips/kernel/irq.c	2021-12-05 23:19:32.892112098 +0200
+++ linux-2.6.36/arch/mips/kernel/irq.c	2010-10-20 23:30:22.000000000 +0300
@@ -156,14 +156,13 @@
  * SMP cross-CPU interrupts have their own specific
  * handlers).
  */
-__IMEM void __irq_entry do_IRQ(int irq)//xflu 20120522
+void __irq_entry do_IRQ(unsigned int irq)
 {
 	irq_enter();
 	__DO_IRQ_SMTC_HOOK(irq);
 	generic_handle_irq(irq);
 	irq_exit();
 }
-EXPORT_SYMBOL(do_IRQ);
 
 #ifdef CONFIG_MIPS_MT_SMTC_IRQAFF
 /*
@@ -171,7 +170,7 @@
  * IRQ affinity, we have this variant that skips the affinity check.
  */
 
-__IMEM void __irq_entry do_IRQ_no_affinity(unsigned int irq)
+void __irq_entry do_IRQ_no_affinity(unsigned int irq)
 {
 	irq_enter();
 	__NO_AFFINITY_IRQ_SMTC_HOOK(irq);
